src/main/java/com/tabnine/selections/AutoImporter.java

```diff
@@ -0,0 +1,170 @@
+import com.intellij.openapi.application.WriteAction;
+import com.intellij.openapi.command.CommandProcessor;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.openapi.editor.ex.EditorEx;
+import com.intellij.openapi.editor.ex.RangeHighlighterEx;
+import com.intellij.openapi.editor.ex.util.EditorUtil;
+import com.intellij.openapi.editor.impl.event.MarkupModelListener;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.Disposer;
+import com.intellij.openapi.util.Key;
+import com.intellij.openapi.util.Pair;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.psi.PsiDocumentManager;
+import com.intellij.psi.PsiFile;
+import com.intellij.util.ObjectUtils;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class AutoImporter implements MarkupModelListener {
+
+  private static final Key<AutoImporter> TABNINE_AUTO_IMPORTER_KEY =
+      Key.create("TABNINE_AUTO_IMPORTER");
+
+  private int startOffset;
+  private int endOffset;
+  @NotNull private final Project project;
+  @NotNull Editor editor;
+  private Disposable myDisposable;
+  private final Map<String, Pair<Integer, Boolean>> possibleRequiredImports = new HashMap<>();
+  private final List<HighlightInfo.IntentionActionDescriptor> importFixes = new ArrayList<>();
+
+  private AutoImporter(@NotNull InsertionContext context) {
+    project = context.getProject();
+    editor = context.getEditor();
+  }
+
+  private void init(@NotNull InsertionContext context) {
+    myDisposable = Disposer.newDisposable();
+    EditorUtil.disposeWithEditor(editor, myDisposable);
+    ((EditorEx) editor).getFilteredDocumentMarkupModel().addMarkupModelListener(myDisposable, this);
+
+    startOffset = context.getStartOffset();
+    endOffset = context.getTailOffset();
+    String insertedText = context.getDocument().getText(new TextRange(startOffset, endOffset));
+    int index = 0;
+    for (String term : insertedText.split("\\W+")) {
+      if (possibleRequiredImports.containsKey(term)) {
+        continue;
+      }
+      possibleRequiredImports.put(term, Pair.create(++index, false));
+    }
+  }
+
+  public static void registerTabNineAutoImporter(
+      @NotNull InsertionContext context) {
+    Editor editor = context.getEditor();
+    AutoImporter autoImporter = editor.getUserData(TABNINE_AUTO_IMPORTER_KEY);
+    if (autoImporter != null) {
+      autoImporter.cleanup();
+      editor.putUserData(TABNINE_AUTO_IMPORTER_KEY, null);
+    }
+    autoImporter = new AutoImporter(context);
+    editor.putUserData(TABNINE_AUTO_IMPORTER_KEY, autoImporter);
+    autoImporter.init(context);
+  }
+
+  @Override
+  public void afterAdded(@NotNull RangeHighlighterEx highlighter) {
+    if (startOffset > highlighter.getAffectedAreaStartOffset()) {
```

Comment:
can all these statements be broken into named functions? it is hard to follow the flow here